File: typescript/tests/compute-catalog-drift.test.ts
/**
 * Drift check vs the Python canonical catalog.
 *
 * Pins the cross-SDK invariant: typescript/src/data/compute_prices.json and
 * python/src/dexcost/data/compute_prices.json are byte-identical, refreshed
 * together by scripts/sync_compute_catalog.sh.
 *
 * Skips gracefully when running from a published npm package (where the
 * Python sibling is not present on disk).
 */

import { describe, expect, test } from "vitest";
import { existsSync, readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const HERE = dirname(fileURLToPath(import.meta.url));
const TS_PATH = join(HERE, "../src/data/compute_prices.json");
// Repo layout: <repo>/typescript/tests/X.test.ts
//              <repo>/python/src/dexcost/data/compute_prices.json
const PY_PATH = join(HERE, "../../python/src/dexcost/data/compute_prices.json");

describe("compute catalog drift", () => {
  test("TS catalog is byte-equal to the Python canonical", () => {
    if (!existsSync(PY_PATH)) {
      // Published-package run — Python sibling not on disk. Skip gracefully.
      // eslint-disable-next-line no-console
      console.warn(
        `[drift-check] Python canonical not found at ${PY_PATH}; skipping byte-equality assertion (this is expected when running from a published npm package).`,
      );
      return;
    }
    const tsBuf = readFileSync(TS_PATH);
    const pyBuf = readFileSync(PY_PATH);
    expect(
      tsBuf.equals(pyBuf),
      `typescript/src/data/compute_prices.json drifted from python/src/dexcost/data/compute_prices.json. ` +
        `Run scripts/sync_compute_catalog.sh to re-sync.`,
    ).toBe(true);
  });
});

File: rust/tests/compute_catalog_drift_test.rs
//! Drift check vs Python canonical catalog.
//!
//! The Rust SDK's `rust/src/data/compute_prices.json` is synced from Python's
//! canonical `python/src/dexcost/data/compute_prices.json` via
//! `scripts/sync_compute_catalog.sh`. This test asserts byte-equality so a
//! drift is caught in CI rather than at runtime.
//!
//! Skip gracefully in published-crate environment where the Python file isn't
//! reachable (e.g. when building from a sdist / crates.io tarball).

use std::path::PathBuf;

#[test]
fn rust_catalog_matches_python_canonical_byte_equal() {
    // Both files live at well-known paths in the monorepo. Resolve from
    // CARGO_MANIFEST_DIR (rust/) up to the repo root.
    let rust_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let repo_root = rust_dir
        .parent()
        .expect("rust/ has a parent (repo root)")
        .to_path_buf();
    let rust_catalog = rust_dir.join("src").join("data").join("compute_prices.json");
    let python_catalog = repo_root
        .join("python")
        .join("src")
        .join("dexcost")
        .join("data")
        .join("compute_prices.json");

    if !python_catalog.exists() {
        eprintln!(
            "[soft-skip] Python canonical catalog not present at {:?} — \
             likely a published-crate build. Skipping drift check.",
            python_catalog
        );
        return;
    }

    let rust_bytes = std::fs::read(&rust_catalog).expect("read rust catalog");
    let python_bytes = std::fs::read(&python_catalog).expect("read python catalog");

    assert_eq!(
        rust_bytes,
        python_bytes,
        "compute_prices.json drift between rust and python — re-run \
         scripts/sync_compute_catalog.sh to re-sync (rust={} bytes, \
         python={} bytes)",
        rust_bytes.len(),
        python_bytes.len(),
    );
}
